### DISCUSSION TITLE
Reload after upgrade

### DISCUSSION THREAD

**Discussion by alexdelprete** [PRProd/HA-Firemote#74] [Q&A]
@PRProd Hi Doug,

every time I upgrade Firemote, and I want to try new functionalities, I don't see them because the frontend is probably caching the resources. I even tried restarting HA but it doesn't help (really strange).

Obviously I tried hard-refresh of the browser but it didn't help.

What is the correct procedure after upgrades to make sure HA is loading latest resources?

Thanks a lot.

**Comment by alexdelprete** [comment]
I managed to solve this by redownloading latest version via HACS, to force this to pop-up:

![image](https://user-images.githubusercontent.com/7027842/232034652-9acdfbb5-80f4-4de4-945c-8641b93ff7ec.png)

After clicking reload, I can see new styles etc.

What service is that reload function calling?

**Comment by PRProd** [comment]
Excellent question @alexdelprete!

The more I research this issue, the more I find that there doesn't seem to be a perfect solution yet.  HA's cache when it comes to .js files is very very sticky.  Even using the standard browser controls to clear the browser cache doesn't always work, which can be a pain.  Since I'm a developer, I have a constant need to reload the latest resource as I'm building it so I've come up with a 95% reliable way to do this in chromium.


## Chrome / Chromium
First, I open the chromium developer tools.  This can be done one of three ways:
 1. Right-click anywhere in your browser window, then scrolldown to "Inspect"
 2. Click the three dots in the upper right hand corner, choose "More Tools" and then "Developer Tools" 
 3. Hold down Ctrl + Shift and press I (i for inspect)
 
![image](https://user-images.githubusercontent.com/25486599/232095972-c692d2fc-06ac-493c-8f0c-864c9ddb849f.png)
![image](https://user-images.githubusercontent.com/25486599/232096262-619d51a7-45dd-4447-ace4-1091bb8cf88e.png)
<br><br>
With the developer tools open, open the Network tab and check the box marked "Disable cache"

![image](https://user-images.githubusercontent.com/25486599/232097746-d967ae49-81b8-4dcf-b64c-5dded5eb2550.png)
<br><br>

At this point, If you press F5 or click your browser's refresh button, you will see the the developer tools screen fill with information related to your page re-load.  If you do not see a console window at the bottom of the developer tools, pressing the ESC key will likely toggle it back into view.

![image](https://user-images.githubusercontent.com/25486599/232099397-19d2de34-9268-4f13-a627-8b6566239181.png)
<br><br>

In the console window, you might find some information that could be helpful for you in the future.  For instance, if you ask yourself "Do I actually have Firemote version 2.1.6 installed right now?" you could find that answer in the console window:

![image](https://user-images.githubusercontent.com/25486599/232099901-e7c590f7-73bf-4112-820b-b6f8cb3bad13.png)
<br><br>

These next steps I do to ensure a full proper refresh are quite heavy handed, and it's possible that not all of these steps are required.  I have found that in my setup, this seems to work almost all of the time so it's what I do.

Click on the clear buttons in both the network tab and the console.

![image](https://user-images.githubusercontent.com/25486599/232101121-b6256824-b331-4af5-ab2c-54a9f628e322.png)
![image](https://user-images.githubusercontent.com/25486599/232100813-dbe659ef-8811-4b1b-80c1-b74171045adf.png)
![image](https://user-images.githubusercontent.com/25486599/232101284-777b1ff2-9b41-44fd-83bc-c130dcb94010.png)
<br><br>

Next, I hold down **CTRL + ALT + SHIFT** and then click the browser's reload button
![image](https://user-images.githubusercontent.com/25486599/232101838-336f212d-6c37-4598-a778-8689254af0ad.png)

This seems to be the magic process that works.  It may not be appropriate in every situation, but it works for me.  
<br>
<br>

## iOS Home Assistant App
First, click on "Settings" in your Home Assistant App

Then, select Companion App
![image](https://user-images.githubusercontent.com/25486599/232105189-b2e3dfce-d782-488b-81a3-626760e6dbb1.png)
<br><br>
Next, select Debugging:
![image](https://user-images.githubusercontent.com/25486599/232105467-04ff04b9-595d-4d00-99c5-c85ef7a424a8.png)
<br><br>
Then finally, click on Reset Frontend Cache
![image](https://user-images.githubusercontent.com/25486599/232105678-6a1fc681-3172-4715-9bf6-081a4a736acb.png)

This seems to work most of the time.  If it doesn't work right away, I suggest closing the Home Assistant app and re-opening it.  If it's still not updated, try the above steps again.
<br>
<br>

## Android Home Assistant App
This one is the trickiest for me, I usually have to repeat these steps 2 or 3 times before I get anywhere.  Hopefully I will someday find a better way, but as of today this is what I do.

Swipe the screen from top to bottom to access the pulldown menu.  Click the gear icon to access the settings menu
![image](https://user-images.githubusercontent.com/25486599/232108113-75f1d3be-0182-4fc1-8833-f26be9710364.png)
<br><br>

Next, locate and click on the Apps menu
![image](https://user-images.githubusercontent.com/25486599/232108626-1e6908c2-2b48-4f18-9055-408e94235d32.png)
<br><br>

Then, scroll through your apps until you find the Home Assistant app, then click on it
![image](https://user-images.githubusercontent.com/25486599/232109455-56d21b50-1f04-4e59-9fe4-2e87ee3b24c5.png)
<br><br>

The App Info menu should be open, showing the details for the currently installed Home Assistant App.  Find "Storage" and click on it
![image](https://user-images.githubusercontent.com/25486599/232110537-90eaa785-8659-4b9a-99b7-e177652c3071.png)
<br><br>

You should now see the size of the current app cache, and a button to clear the app's cache.  Click on the "Clear cache" button
![image](https://user-images.githubusercontent.com/25486599/232112193-1fff77b5-227a-45d4-9c5d-48baa206f002.png)
<br><br>

Here is where it gets interesting / frustrating... the screen shows that the cache has been cleared
![image](https://user-images.githubusercontent.com/25486599/232114412-2b3b82fb-7dfc-48dd-9678-1725349e00bf.png)
<br><br>

BUT... if you click back one menu, and step back into this one, you might find this
![image](https://user-images.githubusercontent.com/25486599/232115083-2be2eafa-7031-4c5d-87d2-60952762e4b0.png)
<br><br>

The cache has quickly grown back from zero.  It's nowhere near the size it was when we first got here, but it's not empty either.  I'm uncertain if this behavior is the reason why it sometimes takes multiple tries, a Force Stop, or a device reboot to ultimately show me the updated content, but either way it can be very frustrating.
<br><br>

# Conclusion
I wish there was a better way right now to handle this annoyance.  If you (or anyone else reading this) happen to have a better solution, please share it!  I did find [one thread online](https://community.home-assistant.io/t/ha-caching/313906) where this problem was being discussed by people with authority, but it hasn't seen any action for several months.

As far as replicating what happens when you click the RELOAD button on the Home Assistant dialog that you posted, I will need to look into that further to see if there is anything to learn from it.

I hope this helps!

**Comment by alexdelprete** [comment]
Wow Doug, I didn't want you to spend all that time writing that post. Sorry for that, but I much appreciate your kindness.

Actually the method in my second post of this thread is working fine. I'd love to know what that RELOAD button is actually doing so we can come up with some script and a button in lovelace that calls it.

And I know that this can be done because when I started with HA a couple of years ago, I found on the forum a lovelace card with 3 buttons, one of which was lovelace reload, that I think was doing the same thing, but at the time I felt I didn't needed, so I deleted it, and now I can't find the post with that code anymore.

Anyway, for me the simplest solution when upgrading will be to redownload firemote and hit reload after the installation, when asked to do so. :)

Thanks again, you are a very kind person, and thanks for your work.

**Comment by PRProd** [comment]
I wasn't sure how much you wanted to know.  :smile: 

Is this the integration you were talking about?  https://github.com/mathoudebine/homeassistant-browser-control-card

**Comment by alexdelprete** [comment]
No it was some custom code a user posted of his card, and there were these 3 buttons.

**Comment by PRProd** [comment]
@alexdelprete ~ Fun fact: I took this thread and made it into a [Firemote Wiki](https://github.com/PRProd/HA-Firemote/wiki/) article called [Force a Refresh](https://github.com/PRProd/HA-Firemote/wiki/Force-a-Refresh), and it has become one of the most popular pages on the Wiki.

We are not alone!  :smile:

**Reply by alexdelprete** [reply]
I found a quick way to do it. I'll try to make a short video.

UPDATE: I forgot I already mentioned it in [second post](https://github.com/PRProd/HA-Firemote/discussions/74#discussioncomment-5615163) of this thread. :)

No need for the video, it's simple, but I didn't see it mentioned in the wiki. ;)
